### ISSUE TITLE
Support non-standard (test-)requirements.txt files

### ISSUE THREAD

**Issue by FlorianWilhelm** [blue-yonder/pyscaffold#79]
Right now you have to use the default requirement file `requirements.txt` and `test-requirements.txt` (since v2.5) to define your requirements and test requirements. Some users complained that they want to have an easy way to define other names and locations of these files. PBR supports right now only to change the name `requirements.txt` by using the environment variable `PBR_REQUIREMENTS_FILES`. A bug request to have `PBR_TEST_REQUIREMENTS_FILES` was filed under https://bugs.launchpad.net/pbr/+bug/1510890. When PBR supports this, PyScaffold will add it as well as a comment in the template of `setup.py`.

**Comment by HolgerPeters** [comment]
What I rather would like to see is a way to separate the dependencies of a project and the environment that a project is run under. I.e. when the dependency is scikit-learn, a build environment would be scikit-learn==0.17.0 and numpy-1.11.0. Basically pinning your environment is the only pratical way of writing software when software is pulled from pypi. I am more or less adapting any project that I created with pyscaffold to use https://github.com/nvie/pip-tools 

I am not so much in favor of test requirement files, if they fill the setuptools  test requirements fields. Why? The issue is that dependencies that are installed as eggs in the `.eggs` directory often don't work as expected. I.e. I would rephrase the issue with test requirements as

* As a python module developer, I would like to separate dependencies for testing and building my project, from dependencies that my module needs when being installed.
* As a python module developer, I would like release distributables of my module with only loose dependencies, so that consumers of my library are free to liberally combine modules without being constrained by the versions I used in my tests.
* As a python module developer, I would like to have a fine control over the environment that I test my software with, so that I am not caught up in a situation where my build breaks, just because one of my dependencies has published a new distributable on pypi.

So:

* conceptionally we should separate dependencies from environments: a package needs both.
* virtual environments are cheap: we should not use "hacks" like "test requirements", rather use different virtual environments.

**Comment by FlorianWilhelm** [comment]
I haven't used pip-tools so far. That means you have unpinned versions when building the package for PyPI but keep them pinned in your virtual environment? Sounds reasonable. How do we handle the fact that PBR reads the ``requirements.txt`` file when running ``python setup.py sdist bdist_wheel``. If I get pip-tools right, unpinned versions are kept in ``requirements.in`` and pinned ones in ``requirements.txt``.

Regarding your other point. If we want to drop the ``test-requirements.txt`` file, we should start with a ``v1.6`` branch and make sure that if a user has not installed py.test and pytest-cov he/she gets a warning when running ``python setup.py test``. I would support that since you are right about the fact that eggs are causing trouble more often than not. 

You wanna go for those two changes?

**Comment by HolgerPeters** [comment]
here is how I would implement this: I would use requirements.txt and test-requirements.txt as inputs for pip-compile. The requirements-txt file resulting from pip-compile would then be named something like `build_environment.txt`. This is the file I would use with `pip-sync` to provision my virtual environment. The `build_environment.txt` could be included into the documentation of the project by a sphinx `literalinclude` directive, so users can see with what kind of an environment the tests were executed.

Would this in any way violate / get into conflict with PBR?

**Comment by FlorianWilhelm** [comment]
@HolgerPeters That sounds like a plan. As long as ``requirements.txt`` contains the unpinned version that should be fine with PBR. This approach would even allow our users to use this functionality optionally. Should we completely get rid of ``test-requirements.txt`` then? We should start a ``v2.6.X`` branch to do all that.

**Comment by FlorianWilhelm** [comment]
Since PBR will be dropped with version 3 of PyScaffold, no `requirements.txt` will be read by `setup.py`. Instead requirements will be directly defined in setup.cfg as specified by newer versions of setuptools.
